lambdas/drivingpermitcheck/src/main/java/uk/gov/di/ipv/cri/drivingpermit/api/service/dvla/endpoints/TokenRequestService.java

```diff
@@ -0,0 +1,269 @@
+import static uk.gov.di.ipv.cri.drivingpermit.library.metrics.ThirdPartyAPIEndpointMetric.DVLA_TOKEN_RESPONSE_TYPE_VALID;
+
+public class TokenRequestService {
+
+    private static final Logger LOGGER = LogManager.getLogger();
+
+    private static final String ENDPOINT_NAME = "token endpoint";
+
+    private final String tokenTableName;
+    private final DynamoDbTable<TokenItem> tokenTable;
+
+    private final URI requestURI;
+    private final String authValue;
+
+    private final HttpRetryer httpRetryer;
+    private final RequestConfig requestConfig;
+
+    private final ObjectMapper objectMapper;
+    private final EventProbe eventProbe;
+
+    // Token item shared between concurrent lambdas (if scaling)
+    public static final String TOKEN_ITEM_ID = "TokenKey";
+    public static final Key TOKEN_ITEM_KEY = Key.builder().partitionValue(TOKEN_ITEM_ID).build();
+
+    // DynamoDB auto ttl deletion is the best effort (upto 48hrs later...)
+    // Token Item ttl expiration enforced CRI side (vs dynamo filter expression)
+    // as there will be only ever be one token.
+    private static final long MAX_ALLOWED_TOKEN_LIFETIME_SECONDS = 3600L;
+    private static final long TOKEN_EXPIRATION_WINDOW_SECONDS = 300L;
+    private static final long TOKEN_ITEM_TTL_SECS =
+            MAX_ALLOWED_TOKEN_LIFETIME_SECONDS - TOKEN_EXPIRATION_WINDOW_SECONDS;
+
+    public static final String INVALID_EXPIRY_WINDOW_ERROR_MESSAGE =
+            "Token expiry window not valid";
+
+    public TokenRequestService(
+            DvlaConfiguration dvlaConfiguration,
+            DynamoDbEnhancedClient dynamoDbEnhancedClient,
+            HttpRetryer httpRetryer,
+            RequestConfig requestConfig,
+            ObjectMapper objectMapper,
+            EventProbe eventProbe) {
+
+        // Token Table
+        tokenTableName = dvlaConfiguration.getTokenTableName();
+        tokenTable =
+                dynamoDbEnhancedClient.table(tokenTableName, TableSchema.fromBean(TokenItem.class));
+
+        this.requestURI = URI.create(dvlaConfiguration.getTokenEndpoint());
+
+        this.httpRetryer = httpRetryer;
+        this.requestConfig = requestConfig;
+
+        this.objectMapper = objectMapper;
+        this.eventProbe = eventProbe;
+
+        // Basic Auth = "username:password" in base64
+        String usernameAndPassword =
+                String.format(
+                        "%s:%s", dvlaConfiguration.getUsername(), dvlaConfiguration.getPassword());
+        String usernameAndPasswordHash =
+                Base64.getEncoder().encodeToString(usernameAndPassword.getBytes());
+        authValue = String.format("Basic %s", usernameAndPasswordHash);
+    }
+
+    public String requestAccessToken(boolean alwaysRequestNewToken)
+            throws OAuthErrorResponseException {
+
+        LOGGER.info("Checking Table {} for existing cached token", tokenTableName);
+
+        TokenItem tokenItem = getTokenItemFromTable();
```

Comment:
Is this there a reason were not using the cri lib data store here?

Reply:
The basic operations of dynamo get/put are all that are used, so I decided against including a second class to just wrap around get/put.

The token service does not do an get then update, it uses put to overwrite (replacing) the token that is there with a fresh one.

Reply:
The downside of this is if we ever want to change our data supplier this would be missed. The data store interface offers some flexibility